### PR TITLE
chore: Apply Bitbucket SAAS Oauth flow

### PR DESCRIPTION
**Pull request by vinokurig**
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

### What does this PR do?
Since we support only Bitbucket Server Oauth flow, added a support for the SAAS version

depends on https://github.com/eclipse-che/che-operator/pull/1445
### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/21517

### How to test this PR?

1. Deploy Eclipse che with the PR image and specific operator image: `chectl server:deploy -p=openshift --che-operator-image=quay.io/ivinokur/che-operator:next -i=quay.io/ivinokur/che-server:next`
2. Create an oauth application in your Bitbucket account: https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/
Set `<eclipse-che endpoint>/api/oauth/callback` as a `Callback URL`
Set all `Account` and `Repository` permissions
3. Create bitbucket oauth secret in the che-server namespace:
```yaml
kind: Secret
apiVersion: v1
metadata:
  name: bitbucket-oauth-config
  namespace: eclipse-che
  labels:
    app.kubernetes.io/component: oauth-scm-configuration
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:
    che.eclipse.org/oauth-scm-server: bitbucket
data:
  id:<base 64 encoded client id from the bitbucket oauth consummer>
  secret: <base 64 encoded secret from the bitbucket oauth consummer>
type: Opaque
```
4. Try to create a factory from a private bitbucket repository.
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      String clientId, String clientSecret, String[] redirectUris, String authUri, String tokenUri)
+      throws IOException {
+    configure(
+        clientId, clientSecret, redirectUris, authUri, tokenUri, new MemoryDataStoreFactory());
+  }
+
+  @Override
+  public User getUser(OAuthToken accessToken) throws OAuthAuthenticationException {
+    BitbucketUser user =
+        getJson("https://api.bitbucket.org/2.0/user", accessToken.getToken(), BitbucketUser.class);
```
Comment (svor) [inline wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java:38]: Maybe it's better to have a constant for `https://api.bitbucket.org/2.0/user`

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    } catch (IOException | InterruptedException | UncheckedIOException e) {
+      throw new ScmCommunicationException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Checks if the provided url belongs to this client (Bitbucket)
+   *
+   * @param scmServerUrl the SCM url to verify
+   * @return If the provided url is recognized by the current client
```
Comment (svor) [inline wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java:212]: @return {@code true} if the provided url is recognized by the current client

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketFactoryParametersResolver.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import org.eclipse.che.api.factory.shared.dto.FactoryVisitor;
+import org.eclipse.che.api.factory.shared.dto.ScmInfoDto;
+import org.eclipse.che.api.workspace.server.devfile.URLFetcher;
+import org.eclipse.che.api.workspace.shared.dto.ProjectConfigDto;
+import org.eclipse.che.api.workspace.shared.dto.devfile.ProjectDto;
+
+/**
+ * Provides Factory Parameters resolver for bitbucket repositories.
+ *
+ * @author Florent Benoit
```
Comment (svor) [inline wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketFactoryParametersResolver.java:41]: Should it be updated?

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketSourceStorageBuilder.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.Map;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.shared.dto.ProjectConfigDto;
+import org.eclipse.che.api.workspace.shared.dto.SourceStorageDto;
+import org.eclipse.che.api.workspace.shared.dto.devfile.SourceDto;
+
+/**
+ * Create {@link ProjectConfigDto} object from objects
+ *
+ * @author Florent Benoit
```
Comment (svor) [inline wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketSourceStorageBuilder.java:27]: Should is be updated?

**Review by svor — APPROVED** [review]
Added a few small remarks. 
Tested it on the OS 4.10 and it works:
- I was redirected to the Bitbucket authentication: 
screenshot-bitbucket org-2022 07 25-15_24_14
- As a result private project was cloned:
screenshot-eclipse-che apps ci-ln-hlz9d1t-72292 origin-ci-int-gce dev rhcloud com-2022 07 25-14_41_38

One strange behavior:
If set all permissions in the Bitbucket application:
screenshot-bitbucket org-2022 07 25-14_53_31
Workspace won't start:
screenshot-eclipse-che apps ci-ln-hlz9d1t-72292 origin-ci-int-gce dev rhcloud com-2022 07 25-14_55_06

But i don't think it's related to the current changes

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketFactoryParametersResolver.java (hunk 3 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketFactoryParametersResolver.java]: fixed

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketSourceStorageBuilder.java (hunk 4 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketSourceStorageBuilder.java]: fixed

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 5]
```diff
@@ -231,6 +231,22 @@ che.oauth1.bitbucket.privatekeypath=NULL
...
 che.oauth1.bitbucket.endpoint=NULL
 
+# Configuration of Bitbucket OAuth2 client. Used to obtain Personal access tokens.
+# Location of the file with Bitbucket client id.
+che.oauth2.bitbucket.clientid_filepath=NULL
+
+# Location of the file with Bitbucket client secret.
+che.oauth2.bitbucket.clientsecret_filepath=NULL
+
+# BitBucket OAuth authorization URI.
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: fixed

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java (hunk 1 above)
Comment (vinokurig) [reply wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java]: done

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java (hunk 2 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java]: done

**Comment by vinokurig** [comment]
@svor 
> Just one strange behavior:
If set all permissions in the Bitbucket application:

Thank you for the catch, I will clarify this in the documentation part.

**Review by tolusha — APPROVED** [review]

**Pull request merged**